numpy/ma/core.py

```diff
@@ -2785,10 +2785,12 @@ def __new__(cls, data=None, mask=nomask, dtype=None, copy=False,
         if isinstance(data, MaskedArray) and (data.shape != _data.shape):
             copy = True
         # Careful, cls might not always be MaskedArray.
-        if not isinstance(data, cls) or not subok:
-            _data = ndarray.view(_data, cls)
-        else:
+        # we must never do .view(MaskedConstant), as that would create a new
+        # instance of np.ma.masked, which would be confusing
+        if isinstance(data, cls) and subok and not isinstance(data, MaskedConstant):
             _data = ndarray.view(_data, type(data))
```

Comment:
Not too sure either. All I did here was swap the order of the if in order to make it easier to see _what_ is happening.

I think the reason might be to ensure that a copy of the wrapper is made - as otherwise we'd modify properties of the input array